### PR TITLE
V2 site overwrite magic __call method

### PR DESCRIPTION
**Pull request by Levdbas**
Related:

- [Overwrite __call method](https://github.com/timber/timber/issues/2741#issuecomment-1713435871_)
- https://github.com/timber/timber/pull/1700

## Issue
When calling a site option that does not have it's own method, like `site.home`. Timber would use the magic `__call` method from the Core class. This magic method by defaults calls the `meta()` method. Since the `meta()` method was deprecated in the site class (https://github.com/timber/timber/pull/1700) the magic `__call `method throws a deprecation notice.

## Solution
Overwrite the __call method in Site.php and call the appropriate `option()` method instead.

## Impact
Less questions on why deprecation notices are thrown.

## Usage Changes
None.

## Considerations
Sometimes it's hard to wrap your head around what are actual instantiated data in Timber and what is called via these magic methods. I always thought that `{{ site.home }}` was a valid method, but when digging deeper I found out that you should call` {{ site.link }}`

I would be in favor of deprecating these magic methods in the future as well so people (including myself) will learn the difference.

## Testing
No, should it be tested?

### CODE REVIEW

**Comment by Levdbas** [comment]
@gchtr , Directly created a PR for this. Should we add this to the 2.0 release?

**Comment by jrathert** [comment]
> Sometimes it's hard to wrap your head around what are actual instantiated data in Timber and what is called via these magic methods. [...] I would be in favor of deprecating these magic methods in the future as well so people (including myself) will learn the difference.

I fully support that.

File: src/Site.php [hunk 1]
```diff
@@ -173,6 +173,39 @@ public function __construct($site_name_or_id = null)
...
+     *
+     * @param string $option     The name of the method being called.
+     * @param array  $arguments Enumerated array containing the parameters passed to the function.
+     *                          Not used.
+     *
+     * @return mixed The value of the option field named `$field` if truthy, `false` otherwise.
+     */
+    public function __call($option, $arguments)
+    {
+        if (\method_exists($this, 'option') && $meta_value = $this->option($option)) {
```
Comment (nlemoine) [inline src/Site.php:202]: We should probably return the value without evaluating it?
```php
return $this->option($option);
```

Comment (Levdbas) [reply src/Site.php]: Ageed, fixed in https://github.com/timber/timber/pull/2798/commits/615917236bef098079189ab9fcf921b8e1eac0f9

**Comment by gchtr** [comment]
> @gchtr , Directly created a PR for this. Should we add this to the 2.0 release?

Thanks for the PR. Yes, because this actually fixes a bug I’d say we should include it in the 2.0 release.



> ## Testing
> No, should it be tested?

@Levdbas I think it would be good if we could add a test for this. Do you need help with that?

**Comment by Levdbas** [comment]
> > @gchtr , Directly created a PR for this. Should we add this to the 2.0 release?
> 
> Thanks for the PR. Yes, because this actually fixes a bug I’d say we should include it in the 2.0 release.
> 
> > ## Testing
> > No, should it be tested?
> 
> @Levdbas I think it would be good if we could add a test for this. Do you need help with that?

Alright I created my first test and used the timber compile_string method for that. removing the __call overwrite in the site class causes the test to fail, so I think we are good here. Any feedback on how to improve this test @gchtr ?

**Review by nlemoine — APPROVED** [review]
Thank you @Levdbas! This looks good to me and congrats on writing test 🙂

**Review by gchtr — APPROVED** [review]
This all looks fine, including the test 👍. Thanks, @Levdbas

**Pull request merged**